projects/output-and-error-handling/README.md

```diff
@@ -0,0 +1,195 @@
+1. Propagating the error to the calling function (and possibly wrapping it with some extra contextual information).
+1. Working around the error to recover from it.
+1. Terminating the program completely.
+1. Ignoring the error - sometimes an error actually doesn't matter.
+
+When we should do each of these isn't always obvious, but here are some guidelines:
+
+#### Propagating the error to the calling function
+
+This is generally our default behaviour. If an error has happened, and we don't know how to handle it, we should early-return from our function, handing the error to the caller.
+
+This means that generally any time we write a function, and it calls another function which may return an error, our function will probably also possibly return an error.
+
+Often times, we want to wrap the error to provide more context. For instance, say we have the following code:
+
+```go
+package main
+
+import "os"
+
+func main() {
+    password, err := readPassword()
+    // ...
+}
+
+func readPassword() (string, error) {
+    password, err := os.ReadFile(".some-file")
+    if err != nil {
+        return "", err
+    }
+    return string(password), nil
+}
+```
+
+If the password file doesn't exist, the error message `open .some-file: no such file or directory` is less useful than an error message like `failed to read password file: open .some-file: no such file or directory`. By wrapping the error with more contextual information, you help the person seeing the error understand _what_ went wrong, _why_ it failed, and what they need to do to fix the situation.
+
+Accordingly, we may write `readPassword` instead like:
+
+```go
+func readPassword() (string, error) {
+    password, err := os.ReadFile(".some-file")
+    if err != nil {
+        return "", fmt.Errorf("failed to read password file: %w", err)
+    }
+    return string(password), nil
+}
+```
+
+You can learn more about creating and wrapping errors in Go by reading [this article](https://earthly.dev/blog/golang-errors/).
+
+#### Working around the error to recover from it
+
+Sometimes, an error may be expected, or may be recoverable. For instance, suppose we have some expensive computation we want to do, but which may have already been done and saved to a file. We may try to read the file, but if we encounter an error that it doesn't exist, we may know how to compute the answer we need instead.
+
+This kind of behaviour will depend entirely on the problem domain you're solving, and there isn't really a general rule for when it's appropriate.
+
+#### Terminating the program completely
+
+A lot of the time, when we run into errors, there's nothing we can do about them.
+
+If we're running a server, and the error happened when processing one request, normally we don't want to terminate our program - we just want to respond to that request saying an error happened, but keep trying to process other requests.
```

Comment:
I'd explicitly mention why here: if you terminate your program because of input from users, you are implicitly giving users the ability to reduce your service's capacity (at least temporarily). This is dangerous.